### PR TITLE
(#2122499) ci: replace LGTM with CodeQL

### PR DESCRIPTION
**Pull request by mrc0mmand**
As LGTM is going to be shut down by EOY, let's use CodeQL instead.

This is loosely based on upstream's CodeQL configs with some minor tweaks to avoid backporting tons of unrelated commits.

rhel-only
Related: #2122499

### CODE REVIEW

**Comment by mrc0mmand** [comment]
With the first three commits I can also finally build RHEL8 systemd on my Fedora machine, which is also nice (but maybe a bit too late given we're already working on 8.8, eh).

As for the CodeQL stuff - I have to build it without gnu-efi, since our systemd doesn't build with Ubuntu Jammy's gnu-efi (3.0.13)* - I suspect this is caused by missing 95ba433a5f34baf92921fb58051bc8241f908c0e and similar commits, which added guards around the problematic definitions. But since we don't use this stuff in RHEL anyway, I don't think it's worth the time messing around with backports (especially since we miss quite a lot of them in that specific area).

Other than that it seems to be doing *The Right Thing™* - *344 new alerts including 117 critical severity security vulnerabilities*.

As usual, the first CodeQL run will fail, since it found new alerts, subsequent runs will pass unless it finds something new.

<details>
<summary>* gnu-efi build error</summary>

```
   FAILED: src/boot/efi/console.c.o 
  /usr/bin/gcc -c ../src/boot/efi/console.c -o src/boot/efi/console.c.o -Wall -Wextra -std=gnu90 -nostdinc -ggdb -O0 -fpic -fshort-wchar -ffreestanding -fno-strict-aliasing -fno-stack-protector -Wsign-compare -Wno-missing-field-initializers -isystem /usr/include/efi -isystem /usr/include/efi/x86_64 -include src/boot/efi/efi_config.h -mno-red-zone -mno-sse -mno-mmx -DEFI_FUNCTION_WRAPPER -DGNU_EFI_USE_MS_ABI
  ../src/boot/efi/console.c:15: warning: "EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL_GUID" redefined
     15 | #define EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL_GUID \
        | 
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:22: note: this is the location of the previous definition
     22 | #define EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL_GUID \
        | 
  ../src/boot/efi/console.c:30:3: error: conflicting types for ‘EFI_KEY_STATE’; have ‘struct <anonymous>’
     30 | } EFI_KEY_STATE;
        |   ^~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:32:3: note: previous declaration of ‘EFI_KEY_STATE’ with type ‘EFI_KEY_STATE’
     32 | } EFI_KEY_STATE;
        |   ^~~~~~~~~~~~~
  ../src/boot/efi/console.c:35:3: error: conflicting types for ‘EFI_KEY_DATA’; have ‘struct <anonymous>’
     35 | } EFI_KEY_DATA;
        |   ^~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:37:3: note: previous declaration of ‘EFI_KEY_DATA’ with type ‘EFI_KEY_DATA’
     37 | } EFI_KEY_DATA;
        |   ^~~~~~~~~~~~
  ../src/boot/efi/console.c:37:29: error: conflicting types for ‘EFI_INPUT_READ_KEY_EX’; have ‘EFI_STATUS (__attribute__((ms_abi)) *)(struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *, EFI_KEY_DATA *)’ {aka ‘long unsigned int (__attribute__((ms_abi)) *)(struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *, EFI_KEY_DATA *)’}
     37 | typedef EFI_STATUS (EFIAPI *EFI_INPUT_READ_KEY_EX)(
        |                             ^~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:68:10: note: previous declaration of ‘EFI_INPUT_READ_KEY_EX’ with type ‘EFI_INPUT_READ_KEY_EX’ {aka ‘long unsigned int (__attribute__((ms_abi)) *)(struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *, EFI_KEY_DATA *)’}
     68 | (EFIAPI *EFI_INPUT_READ_KEY_EX) (
        |          ^~~~~~~~~~~~~~~~~~~~~
  ../src/boot/efi/console.c:47:29: error: conflicting types for ‘EFI_KEY_NOTIFY_FUNCTION’; have ‘EFI_STATUS (__attribute__((ms_abi)) *)(EFI_KEY_DATA *)’ {aka ‘long unsigned int (__attribute__((ms_abi)) *)(EFI_KEY_DATA *)’}
     47 | typedef EFI_STATUS (EFIAPI *EFI_KEY_NOTIFY_FUNCTION)(
        |                             ^~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:82:10: note: previous declaration of ‘EFI_KEY_NOTIFY_FUNCTION’ with type ‘EFI_KEY_NOTIFY_FUNCTION’ {aka ‘long unsigned int (__attribute__((ms_abi)) *)(EFI_KEY_DATA *)’}
     82 | (EFIAPI *EFI_KEY_NOTIFY_FUNCTION) (
        |          ^~~~~~~~~~~~~~~~~~~~~~~
  ../src/boot/efi/console.c:51:29: error: conflicting types for ‘EFI_REGISTER_KEYSTROKE_NOTIFY’; have ‘EFI_STATUS (__attribute__((ms_abi)) *)(struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *, EFI_KEY_DATA,  EFI_STATUS (__attribute__((ms_abi)) *)(EFI_KEY_DATA *), void **)’ {aka ‘long unsigned int (__attribute__((ms_abi)) *)(struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *, EFI_KEY_DATA,  long unsigned int (__attribute__((ms_abi)) *)(EFI_KEY_DATA *), void **)’}
     51 | typedef EFI_STATUS (EFIAPI *EFI_REGISTER_KEYSTROKE_NOTIFY)(
        |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:88:10: note: previous declaration of ‘EFI_REGISTER_KEYSTROKE_NOTIFY’ with type ‘EFI_REGISTER_KEYSTROKE_NOTIFY’ {aka ‘long unsigned int (__attribute__((ms_abi)) *)(struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *, EFI_KEY_DATA *, long unsigned int (__attribute__((ms_abi)) *)(EFI_KEY_DATA *), void **)’}
     88 | (EFIAPI *EFI_REGISTER_KEYSTROKE_NOTIFY) (
        |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../src/boot/efi/console.c:63:16: error: redefinition of ‘struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL’
     63 | typedef struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL {
        |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:102:16: note: originally defined here
    102 | typedef struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL{
        |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../src/boot/efi/console.c:70:3: error: conflicting types for ‘EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL’; have ‘struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL’
     70 | } EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL;
        |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:65,
                   from ../src/boot/efi/console.c:6:
  /usr/include/efi/eficonex.h:109:3: note: previous declaration of ‘EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL’ with type ‘EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL’
    109 | } EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL;
        |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/efi/efi.h:42,
                   from ../src/boot/efi/console.c:6:
  ../src/boot/efi/console.c: In function ‘console_key_read’:
  ../src/boot/efi/console.c:96:74: warning: passing argument 1 of ‘TextInputEx->ReadKeyStrokeEx’ from incompatible pointer type [-Wincompatible-pointer-types]
     96 |                 err = uefi_call_wrapper(TextInputEx->ReadKeyStrokeEx, 2, TextInputEx, &keydata);
        |                                                                          ^~~~~~~~~~~
        |                                                                          |
        |                                                                          EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL * {aka struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *}
  ../src/boot/efi/console.c:96:74: note: expected ‘struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *’ but argument is of type ‘EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *’ {aka ‘struct _EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *’}
```
</details>

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# vi: ts=2 sw=2 et:
+# SPDX-License-Identifier: LGPL-2.1-or-later
+#
+name: "CodeQL"
+
+on:
+  pull_request:
+    branches:
+      - master
+      - rhel-*
```
Comment (jamacku) [inline .github/workflows/codeql.yml:11]: Wouldn'T be better: :smile: 
```suggestion
      - rhel-8.*.0
```

File: .github/workflows/codeql.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    paths:
+      - '**/meson.build'
+      - '.github/**/codeql*'
+      - 'src/**'
+      - 'test/**'
+      - 'tools/**'
+  push:
+    branches:
+      - master
+      - rhel-*
```
Comment (jamacku) [inline .github/workflows/codeql.yml:21]: Same here

**Review by jamacku — COMMENTED** [review]
Just nitpicks, LGTM :+1:

**Review by dtardon — APPROVED** [review]
I trust you that all these .yml files actually do what they are supposed to :-)
LGTM

File: .github/workflows/codeql.yml (hunk 1 above)
Comment (mrc0mmand) [reply .github/workflows/codeql.yml:11]: That's unnecessarily specific for no real gain here, imo.

**Review by jamacku — APPROVED** [review]

**Pull request merged**